File: app/javascript/channels/message_read_channel.js
import consumer from "channels/consumer"

let messageReadSubscription = null;

document.addEventListener('turbo:load', () => {
  const channelElement = document.getElementById('messagesContainer') || document.getElementById('messages-container');
  
  if (channelElement) {
    const channelId = channelElement.dataset.channelId;
    
    if (channelId) {
      // Unsubscribe from previous channel if exists
      if (messageReadSubscription) {
        messageReadSubscription.unsubscribe();
      }
      
      messageReadSubscription = consumer.subscriptions.create(
        { 
          channel: "MessageReadChannel",
          channel_id: channelId
        },
        {
          connected() {
            console.log('[MessageReadChannel] Connected to channel:', channelId);
            // Force update the unread count when entering the channel
            setTimeout(() => {
              console.log('[MessageReadChannel] Forcing unread count update for current channel');
              const unreadCountBadge = document.querySelector(`a[href*="/chat/channel/${channelId}"] span:last-child`);
              if (unreadCountBadge && unreadCountBadge.classList.contains('bg-[#FF6D75]')) {
                console.log('[MessageReadChannel] Found unread badge, clearing it');
                if (window.updateChannelUnreadCount) {
                  window.updateChannelUnreadCount(channelId, 0);
                }
              }
            }, 1000);
          },

          disconnected() {
            console.log('[MessageReadChannel] Disconnected');
          },

          received(data) {
            console.log('[MessageReadChannel] Received:', data);
            
            if (data.action === 'message_read') {
              updateReadStatus(data.message_id, data.readers_count);
            }
          },
          
          markAsRead(messageId) {
            this.perform('mark_as_read', { message_id: messageId });
          }
        }
      );
      
      // Mark messages as read when they appear in viewport
      observeMessagesForReading();
    }
  }
});

function updateReadStatus(messageId, readersCount) {
  const messageEl = document.getElementById(`message-${messageId}`);
  if (!messageEl) return;
  
  // Update read status UI (we'll add this next)
  const readIndicator = messageEl.querySelector('.read-indicator');
  if (readIndicator) {
    readIndicator.textContent = `✓✓ ${readersCount}`;
  }
}

function observeMessagesForReading() {
  const options = {
    root: document.getElementById('messagesContainer') || document.getElementById('messages-container'),
    rootMargin: '0px',
    threshold: 0.5
  };
  
  const observer = new IntersectionObserver((entries) => {
    entries.forEach(entry => {
      if (entry.isIntersecting) {
        const messageId = entry.target.id.replace('message-', '');
        if (messageReadSubscription) {
          messageReadSubscription.markAsRead(messageId);
        }
      }
    });
  }, options);
  
  // Observe all messages
  document.querySelectorAll('[id^="message-"]').forEach(messageEl => {
    observer.observe(messageEl);
  });
}

// Re-observe new messages when they are added
document.addEventListener('turbo:stream-render', () => {
  setTimeout(() => {
    observeMessagesForReading();
  }, 100);
});

File: app/channels/message_read_channel.rb
class MessageReadChannel < ApplicationCable::Channel
  def subscribed
    return unless params[:channel_id].present?
    
    @channel = current_user.channels.find_by(id: params[:channel_id])
    return unless @channel
    
    stream_from "channel_#{@channel.id}_reads"
    
    # Mark all existing messages in the channel as read immediately
    mark_channel_messages_as_read
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end
  
  def mark_as_read(data)
    return unless data['message_id'].present?
    
    message = @channel.messages.find_by(id: data['message_id'])
    return unless message
    
    # Mark message as read
    message.mark_as_read_by(current_user)
    
    # Broadcast read status to other users
    ActionCable.server.broadcast(
      "channel_#{@channel.id}_reads",
      {
        action: 'message_read',
        message_id: message.id,
        user_id: current_user.id,
        readers_count: message.readers.count
      }
    )
    
    # Also broadcast to update unread count for this user
    broadcast_unread_count
  end
  
  private
  
  def mark_channel_messages_as_read
    Rails.logger.info "[MessageReadChannel] Marking all messages as read for user #{current_user.id} in channel #{@channel.id}"
    
    unread_messages = @channel.messages.where.not(id: current_user.read_messages.pluck(:id))
    Rails.logger.info "[MessageReadChannel] Found #{unread_messages.count} unread messages"
    
    unread_messages.find_each do |message|
      message.mark_as_read_by(current_user)
    end
    
    # Broadcast updated unread count immediately
    new_count = @channel.unread_messages_count_for(current_user)
    Rails.logger.info "[MessageReadChannel] New unread count: #{new_count}"
    
    broadcast_unread_count
  end
  
  def broadcast_unread_count
    ActionCable.server.broadcast(
      "user_#{current_user.id}_unread",
      {
        action: 'update_unread_count',
        channel_id: @channel.id,
        unread_count: @channel.unread_messages_count_for(current_user)
      }
    )
  end
end
